### DISCUSSION TITLE
Get Jacobian in UserObject

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#26734] [Q&A Modules: Solid mechanics] [answered]
Hello,

is there a way to get the global DOFxDOF jacobian in a UserObject?

Or do I have to assemble it myself?

**Comment by GiudGiud** [comment]
Hello

We do not recommend assembling it yourself.
You should be able to use the TaggingInterface to retrieve the Jacobian. I dont know which tag you would use

If that does not work directly, the create a new matrix tag and add it to all your kernels. Then you can retrive the Jacobian with this matrix tag. 

Guillaume

**Reply by GiudGiud** [reply]
Actually, from the documentation, I think the `system` tag will be helpful
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html

**Reply by Flolaffel** [reply]
I've noticed the `Tensor Mechanics Master Action` has no `matrix_tags `parameter. Will it still contribute to the `system `matrix by default?

**Reply by Flolaffel** [reply]
Furthermore, I can't seem to get the matrix import working.

`_subproblem.getMatrixTags();`

tells me there are two matrices: SYSTEM with ID=0 and TIME with ID=1. But when I do

`_sys.hasMatrix(0);`

it returns false. And as expected 

`_sys.getMatrix(0) `

is not working either. Am I missing an include or something? What am I doing wrong here?

**Reply by GiudGiud** [reply]
> Will it still contribute to the system matrix by default?

yes

What do you mean missing an include? If it is compiling then you are not missing an include.
When is this object executed? I wonder if you are querying for this matrix too early

**Reply by Flolaffel** [reply]
Ah okay, that's good to know. 

I'm executing said commands in the `initialize` function of an `ElementUserObject` at `timestep_end` `post_aux`.

**Reply by GiudGiud** [reply]
@lindsayad thoughts about who to tag on the tagging system and if you, then how would the matrix not be available then?

**Reply by maxnezdyur** [reply, marked as answer]
In `AdjointSolve.C` the way we get the jacobian matrix is
```
 // Adjoint matrix, solution, and right-hand-side
  auto & matrix = static_cast<ImplicitSystem &>(_nl_forward.system()).get_system_matrix();
  ```

**Reply by lindsayad** [reply]
The Jacobian matrix is only associated with a tag during Jacobian evaluation:
```c++
void
NonlinearSystemBase::computeJacobian(SparseMatrix<Number> & jacobian, const std::set<TagID> & tags)
{
  associateMatrixToTag(jacobian, systemMatrixTag());

  computeJacobianTags(tags);

  disassociateMatrixFromTag(jacobian, systemMatrixTag());
}
```

**Reply by Flolaffel** [reply]
@maxnezdyur It works but the matrix I get is not symmetric. Shouldn't the jacobian be symmetric for a linear elastic tensor mechanics problem?

@lindsayad What is the reason for that?

**Reply by GiudGiud** [reply]
@Flolaffel you'll want to add a new matrix tag for now, unless @maxnezdyur 's solution works for you
If you are going through the action and it wont let you then you ll need to modify it

**Reply by lindsayad** [reply]
> @maxnezdyur It works but the matrix I get is not symmetric. Shouldn't the jacobian be symmetric for a linear elastic tensor mechanics problem?

Do you have Dirichlet boundary conditions?

> @lindsayad What is the reason for that?

I do not know for sure why @fdkong did not make a more permanent association at least in standard nonlinear system problems, but for instance in general eigenproblems there is not one canonical matrix, there are two.

**Reply by Flolaffel** [reply]
> Do you have Dirichlet boundary conditions?

I do. I didn't realize the BCs are built in by only zeroing the rows but not columns. That explains the asymmetry. Zeroing the columns corresponding to fixed DOFs restores the symmetry.

Which means @maxnezdyur's solution does what I was looking for:

`auto & system_matrix = static_cast<ImplicitSystem &>(_sys.system()).get_system_matrix();`
